Add ansible roles view and entity

synkd opened the pull request:
This PR adds views and entities for the Ansible Roles page (located in the webUI at Configure > Roles) and the Ansible Variables page (located in the webUI at Configured > Variables).

This is my first Airgun PR, so I would appreciate any feedback.

Note that there are two widgets in the `NewAnsibleVariableView` view that are not currently functioning properly; I have marked these with comments. However, in the interest of getting these changes merged ahead of 6.10 branching next week, I thought it best to get this up for review in its current state.

mshriver commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,57 @@
+        view.flash.assert_no_error()
+        view.flash.dismiss()
+
+    def count_imported_roles(self):
```
Not sure about naming for this in airgun, but I'd think about two things with this function:

1. It's a property (should be decorated, takes no arguments and doesn't change state)
2. I'd call it `imported_roles_count`, but this is kinda subjective and might go against the grain with other entity function names in airgun.

mshriver commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,57 @@
+    endpoint_path = '/ansible/ansible_roles'
+    VIEW = AnsibleRolesImportView
+
+    def prerequisite(self, *args, **kwargs):
```
You can use `NavigateToSibling` here to flatten this quite a bit. I don't think its in wide use in airgun, you followed the 'regular' pattern here.

```
from navmazing import NavigateToSibling # at the top of the file

# then within the class, where you currently are defining prerequisite as a function

prerequisite = NavigateToSibling('All')
```

The class takes care of resolving the object, and navigating at the right time. 

https://github.com/RedHatQE/navmazing/blob/master/navmazing/__init__.py#L165

mshriver commented on airgun/entities/ansible_variable.py:
```diff
@@ -0,0 +1,67 @@
+
+    VIEW = NewAnsibleVariableView
+
+    def prerequisite(self, *args, **kwargs):
```
Same thing applies to this prerequisite too - `NavigateToSibling` and `NavigateToAttribute` are handy classes to define navigation functions that relate to other destinations on the entity they're registered against.

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,75 @@
+    title = Text("//h1[contains(., text()='Ansible Roles')")
+    import_button = Text("//a[contains(@href, '/ansible_roles/import')]")
+    submit = Button('Submit')
+    total_imported_roles = Text("//span[@class='pagination-pf-items-total']")
```
From the xpath, this looks like the pagination widget's `total_items` attribute:
https://github.com/SatelliteQE/airgun/blob/master/airgun/widgets.py#L1453

You can reference it straight through `view.pagination.total_items` which I generally recommend, but if you want to create a more direct reference to the number of imported roles you could define a small property function.

```
@property
def total_imported_roles(self):
    return self.pagination.total_items
```

synkd replied:
Been trying to figure out how to implement this. I see a couple of tests in Robottelo that use do something similar, such as test_contenthost.py:

https://github.com/SatelliteQE/robottelo/blob/9edf53cb6f30c77e13ab4557b1668033406904c1/tests/foreman/ui/test_contenthost.py#L1321

So it looks like I'll need to add something like the `read_all` method from the content host entity in order to access that data:

https://github.com/SatelliteQE/airgun/blob/e6a29c3c0a46dc8a63c9818477fc733ae3a613fb/airgun/entities/contenthost.py#L31

Is something like this what you had in mind, or are there other possibilities?

mshriver replied:
To access this value through the entity itself, yes you would need an entity function like `read_all` to read the view and return the dict.

The Pagination widget itself can just be directly read, or its `total_items` property can be accessed.

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,75 @@
+
+    @property
+    def is_displayed(self):
+        return self.browser.wait_for_element(
```
Consider leveraging more of the defined widgets to assert the view:

```
return (
    self.title.is_displayed and
    self.import_button.is_displayed and
    self.table.is_displayed
)
```

synkd replied:
I'm making this change but leaving off `self.table.is_displayed` as the table is not present when no roles have been imported.

latran commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,91 @@
+    add_matcher_button = Text("//a[contains(@class, 'add_nested_fields')]")
+    attribute_type = SatSelect("//div[@class='matcher-group']/select[1]")
+    # the below attribute_value widget is not locating the element correctly as currently written
+    attribute_value = Text("//div[@class='matcher-group']/input[1]")
```
These will be a little bit difficult to handle because there could be potential multiple `Attribute type` when you click `Add Matcher`.  One shouldn't give much issue when it's trying to search for the one available field though.  Try `@class='matcher_value'` and see if that works.  if it doesn't, you may need to set the `locator` to that or pass in the `input` one.

synkd replied:
I've been working on this today. Still unable to get the `attribute_value` field working. I've tried the following XPath expressions:
- `TextInput("//div[@class='matcher-group']/input[1]")`
- `TextInput("//input[contains(@class, 'matcher_value')")`
- `TextInput("//input[@class='matcher_value']")`
- several other minor variations on the above.

All of these throw a `NoSuchElementException` when my test reaches the `attribute_value` variable. Weirdly, the current XPath expression in the PR (`Text("//div[@class='matcher-group']/input[1]"`) does throw the same exception. However, `Text` objects do not have a `fill` method, and changing `Text` to `TextInput` brings back the exception. I do not have an explanation for this discrepancy in finding the element.

At this point, I am inclined to relegate tests involving matchers to manual tests for the time being, as this issue is preventing us from automating Ansible roles and variables tests that do not involve matchers. @latran @mshriver I would be interested to hear your perspectives on this idea.

synkd replied:
@mshriver @latran I was able to spend some more time on this today. One of the things that I tried was a much more verbose XPath expression to point to this input field:
`TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[1]/td[1]/div[1]/input[1]")`
I verified in a browser console that this expression does indeed point to the intended text input field. However, my test still throws a `NoSuchElementException` when it attempts to use this input field. For reference, the test I'm using is here [1]. 

I could really use some help figuring out why I'm getting this error with an apparently correct XPath expression. 

[1] https://github.com/synkd/robottelo/blob/cf480c9c8c90905acf4912a04f17d27680e88cee/tests/foreman/ui/test_ansible.py#L78

synkd replied:
Incidentally, using XPath expressions that point to specific table rows is a possible solution to the issue of multiple `attribute_type` variables that Lai inititally raised in this thread. I'm thinking of something like:
```
attribute_types = [
    TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[1]/td[1]/div[1]/input[1]"),
    TextInput("//div[@class='matcher-parent']/fieldset[2]/div[1]/table[1]/tbody[1]/tr[2]/td[1]/div[1]/input[1]"),
]
```
The list could contain a number items equal to the largest number of Ansible Variable matchers needed by a test.

omkarkhatavkar replied:
@synkd I think you need to refactor this code. I think adding a table view makes more sense and attaching that to Main View e.g.

```
class MatcherView(View):
    ROOT = '//h2[text()="Specify Matchers  "]/..'
    add_matcher = Text("//a[contains(@class, 'add_nested_fields')]")
    table = SatTable(
        locator=".//table[@class='table white-header']",
        column_widgets={
            'Attribute type': Select(".//select"),
            'Value': TextInput(".//textarea"),
            'Actions': Text(".//a"),
        },
    ) 
```

synkd replied:
@omkarkhatavkar Thank you for looking at this. Using a table widget does make sense, since that's how that part of the HTML is structured. However, I'm not clear on what the advantage would be to adding a table view, as opposed to adding a table to the existing `NewAnsibleVariableView`. Could you please explain your thinking there? Is it to separate filling the values in the `Specify Matchers` section from filling the data on the rest of the page?

omkarkhatavkar replied:
Added more details implementation here https://github.com/synkd/airgun/pull/1, you can modify this behaviour if you like

synkd replied:
Thanks to some amazing collab from @omkarkhatavkar, this is now working as expected. I need to do some additional testing to verify that Omkar's changes to the `CustomParameter` widget do not affect other tests, and  need to resolve an issue with my delete method in the Ansible Variable entity where the confirmation dialog is not being handled properly. Once these tasks are completed, this PR will be ready for re-review.

synkd replied:
One additional note: I had to recreate separate methods for creating a variable with and without overrides. When the `Override` checkbox is selected but no values are entered into the section of the form enabled by that checkbox, Satellite throws an error when the form is submitted. This does not strike me as a product bug, so I've added separate methods for those two scenarios (with and without overrides) to the entity.

latran reviewed (changes requested):
Minor changes

synkd commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,58 @@
+
+    @property
+    def imported_roles_count(self):
+        view = self.navigate_to(self, 'All')
```
Added.

synkd commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,58 @@
+
+@navigator.register(AnsibleRolesEntity, 'All')
+class ShowAllRoles(NavigateStep):
+
```
Added.

synkd commented on airgun/entities/ansible_variable.py:
```diff
@@ -0,0 +1,67 @@
+        view.flash.assert_no_error()
+        view.flash.dismiss()
+
+    def read_total_variables(self):
```
Added.

synkd commented on airgun/entities/ansible_variable.py:
```diff
@@ -0,0 +1,67 @@
+        return view.total_variables.read()
+
+    def create(self, values):
+        view = self.navigate_to(self, 'New')
```
Added.

synkd commented on airgun/entities/ansible_variable.py:
```diff
@@ -0,0 +1,67 @@
+        view.flash.assert_no_error()
+        view.flash.dismiss()
+
+    def create_with_overrides(self, values):
```
Added.

synkd commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,91 @@
+    required = Checkbox(id='ansible_variable_required')
+    validator_type = SatSelect(id='ansible_variable_validator_type')
+    validator_rule = TextInput(id='ansible_variable_validator_rule')
+    # the below attribute_order widget throws an error when a string value is passed to it
```
`E               ValueError: Cannot locate option with visible text: 'fqdn'. Available options: '', ''`

This element is a bit strange; I don't know that there is a similar one elsewhere in Satellite. It's a `textarea` with a `rows` attribute. In the source HTML, the values are listed as a space-separated string (by default, `fqdn hostgroup os domain`). When the HTML is rendered, the order is preserved, but each value is on a separate row. 

I've tried to change the order by passing in both a list (`'attribute_order': ['domain', 'fqdn', 'hostgroup', 'os']` and a string (\'attribute_order': 'domain fqdn hostgroup os'`), but both of these result in the error above.

omkarkhatavkar replied:
If I understand this problem, this is a time synchronization issue. To solve this you can take the help of the before_fill method.

you can take this as example https://github.com/SatelliteQE/airgun/blob/b822d70e44a6e6d46b3ec4abac527b0c7efc07b0/airgun/views/host.py#L531

synkd replied:
Thank you, @omkarkhatavkar. I will try this and report back on the results.

synkd replied:
@omkarkhatavkar was correct about this. With a `before_fill` method added, this is now functioning as expected.

latran commented:
@synkd Any updates on this and were you able to resolve your issue?

synkd commented:
@latran Thanks for following up. I will revisit this PR tomorrow and take stock.

synkd commented:
I believe this PR is ready for review again. A couple of notes:
- For some reason, `handle_alert` is not functioning properly in the Ansible variables delete method. I have not been able to find any meaningful differences between its implementation here and in other entities. I ran a few tests locally that use `handle_alert`, and it worked as expected in those cases. I've added the confirmation button to the Ansible variables view as a workaround.
- I've added a `preimport_check` to the Ansible roles entity. This can be used in Ansible variables tests to ensure that we don't try to create an Ansible variable before importing any roles to which the variable could be assigned.
- Omkar pointed out to me that we need to ensure that his changes to the `CustomParameter` widget don't affect any existing tests that use this widget. As a smoke test, I locally ran the `tests/foreman/ui/test_domain.py` module, which includes 4 tests that use the `CustomParameter` widget. 3 of these passed and 1 failed. However, the failed test also failed in the most recent CI run with the same error that I encountered, so I don't believe the failure is related to the changes to the widget. There are a few other tests that use this widget, and I'm happy to do additional testing if necessary.

omkarkhatavkar commented:
@synkd can you please rebase your branch and see why the CI showing errors.

synkd commented:
@omkarkhatavkar I had defined the same variable twice in one of the view classes, and Docs Build was failing as a result. The CI checks are passing with that fixed.

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+    table = Table(
+        './/table',
+        column_widgets={
+            '▲ Name	': Text("./a"),
```
I think I would raise this as an issue with the UX team, for testability. They have a gchat channel we can engage on instead of going straight to writing a BZ.

mshriver replied:
In the meantime try using `widgetastic.utils.partial_match` instance for the key here, like `partial_match('Name')`. I'm not confident it will work correctly as a dictionary key, but its worth a try.

synkd replied:
Thanks for pointing me to the UX team's channel. I will try the `partial_match` util as well and update with the results here.

synkd replied:
I could not figure out how to get `partial_match` working as a dictionary key, but I came up with another workaround: as long as we're searching for a specific role to delete, there will only be one row in the search results, so `view.table.row()['Actions'].widget.fill('Delete')` works. This is not optimal but I believe it will suffice until the underlying behavior changes (i.e. the UX team decides to either not sort roles by name on import or wraps the sort character in a separate `span` tag).

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+
+    @property
+    def is_displayed(self):
+        breadcrumb_loaded = self.browser.wait_for_element(self.breadcrumb, exception=False)
```
Is there an exception pattern/loading time that you're concerned about here?

There is `wait_displayed`, which wraps `is_displayed` in a wait_for loop, and we generally discourage doing any waiting in the `is_displayed` implementation.  Page 'readiness' issues should also be dealt with in the browser plugin, not in individual `is_displayed` functions.

synkd replied:
I honestly do not recall why I added the `exception=False` here. My `is_displayed` methods were largely borrowed/stolen from other Airgun views, so I have no issue with modifying them if this implementation is not in line with best practices.

synkd replied:
Removed `wait_for_element`.

mshriver commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,129 @@
+
+
+class MatcherActions(View):
+    """Interface table has Acttibute type column that contains select field and
```
typo here

synkd replied:
addressed

mshriver commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,129 @@
+    matcher_key = Select(".//select")
+    matcher_value = TextInput(locator=".//input[@class='matcher_value']")
+
+    def fill(self, values):
```
This should not be necessary, what's happening here is the standard fill pattern for a view class. 

If you want a custom exception class for this (not just an exception message like you have), then just check the fill dict for keys before calling `super().fill()`

omkarkhatavkar replied:
Yes if the values are coming from the test correctly no need to have this. I was apparently testing something and that added this code. You can add a similar test code that references the fields. 

e.g.  `matcher_section.params.matcher_key`

mshriver commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,129 @@
+
+    @property
+    def is_displayed(self):
+        breadcrumb_loaded = self.browser.wait_for_element(self.breadcrumb, exception=False)
```
Same consideration as the other class about using `wait_for` within is_displayed.

synkd replied:
Removed the `wait_for_element`.

mshriver commented on airgun/widgets.py:
```diff
@@ -916,20 +916,25 @@ class CustomParameter(Table):
     add_new_value = Text("..//a[contains(text(),'+ Add Parameter')]")
 
-    def __init__(self, parent, locator=None, id=None, logger=None):
+    def __init__(self, parent, locator=None, id=None, logger=None, column_widgets=None):
```
`Table` has a lot more kwargs than this, so the function signature here should be updated to take `**kwargs` and pull any references from there, instead of trying to maintain an exact signature with the parent class.

synkd replied:
I've attempted to address this but would appreciate feedback on my implementation.

mshriver commented on airgun/widgets.py:
```diff
@@ -938,9 +943,9 @@ def read(self):
-            parameters.append({'name': name, 'value': value})
+            name = row[self.name].widget.read()
+            value = row[self.value].widget.read()
+            parameters.append({self.name_key: name, self.value_key: value})
```
Are `name_key` and `value_key` doing anything functionally here, other than replacing the use of the string `'name'` and `'value'`?

mshriver replied:
I think it just clicked that all of this is being done to work around the fact that there's a table that is including the sort character.

That doubles my commitment to the idea that we should push back on UX and exclude the sorting indicator from the text content of the column.

omkarkhatavkar replied:
Yes because of the hard coding done on the table name this was not reusable, I was just trying to make that reusable.

mshriver reviewed (changes requested):
Looks like some questionable things in the interest of working around sorting behavior that I'd like us to discuss with dev.

synkd commented on airgun/widgets.py:
```diff
@@ -916,20 +916,25 @@ class CustomParameter(Table):
             ),
         }
+        self.name = list(column_widgets.keys())[0]
+        self.name_key = '_'.join(self.name.lower().split(' '))
```
@mshriver Was there a change that you wanted made relating to this comment? It was not totally clear to me.

mshriver commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,75 @@
+        view = self.navigate_to(self, 'All')
+        view.search(entity_name)
+        view.table.row()['Actions'].widget.fill('Delete')
+        view.confirm.confirm_dialog.click()
```
You should be able to use `view.dialog.confirm()` here, as the PF4 confirmation dialog modal is on the BaseView as `dialog`.

mshriver commented on airgun/entities/ansible_variable.py:
```diff
@@ -0,0 +1,74 @@
+        view = self.navigate_to(self, 'New')
+        view.override.fill(True)
+        view.expand()
+        view.matcher_section.before_fill(values)
```
This is slightly odd, the pattern for before_fill/after_fill is such that they shouldn't have to be called directly.

As `view.fill` executes, it will call `view.matcher_section.fill` with `values['matcher_section']`, which in turn should execute `self.before_fill`.

Not a blocking comment, going to keep reading, as I'm guessing this is trying to enforce fill order.

mshriver commented on airgun/entities/configgroup.py:
```diff
@@ -43,6 +43,7 @@ def delete(self, entity_name):
         view = self.navigate_to(self, 'All')
         view.searchbox.search(entity_name)
         view.table.row(name=entity_name)['Actions'].widget.click(handle_alert=True)
+        view.dialog.confirm()
```
I think you need to rebase against master, this is already in.

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+    is present, without the search widget or table.
+    """
+
+    title = Text("//h1[contains(., text()='Ansible Roles')")
```
non-blocking, but `normalized-space()` is the best way to do this locator.

https://github.com/SatelliteQE/airgun/issues/591

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+    title = Text("//h1[contains(., text()='Ansible Roles')")
+    import_button = Text("//a[contains(@href, '/ansible_roles/import')]")
+    submit = Button('Submit')
+    total_imported_roles = Text("//span[contains(@class, 'pagination-pf-items-total')]")
```
This is actually already available as `self.pagination.total_items`.

https://github.com/SatelliteQE/airgun/blob/master/airgun/widgets.py#L1463

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+    table = Table(
+        './/table',
+        column_widgets={
+            'Name': Text("./a"),
```
I know there's a long history of discussion about this table definition and its behavior.

Are you trying to define column widgets here for every column header? Usually you would only define the column that actually has an interactable widget.

The `Table` widget otherwise has properties/attributes/functions to deal with the table headers/columns, outside of `column_widgets`.

Non-blocking, but its a bit odd to see `Text` widgets being used here, the table cells are just 'normal' text.

mshriver commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,72 @@
+        },
+    )
+    pagination = Pagination()
+    confirm = Pf4ConfirmationDialog()
```
This is in `BaseLoggedInView` as `dialog`, please update to use that widget where necessary.

mshriver commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,116 @@
+
+    title = Text("//h1[contains(., text()='Ansible Variables')")
+    new_variable = Text("//a[contains(@href, '/ansible/ansible_variables/new')]")
+    total_variables = Text("//span[@class='pagination-pf-items-total']")
```
Again, available through `self.pagination`, non-blocking.

mshriver commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,116 @@
+                self.add_matcher.click()
+
+    @property
+    def expand_button(self):
```
doesn't look like this is actually used anywhere. Its also slightly risky naming, as it returns an element instance, not a button widget instance.

mshriver commented on airgun/views/dashboard.py:
```diff
@@ -124,6 +124,11 @@ def fill(self, values):
                 'No. of Tasks'
             ].widget.click()
 
+    @View.nested
```
rebase should remove this from the diff view, its already in.

mshriver commented on airgun/widgets.py:
```diff
@@ -916,31 +916,44 @@ class CustomParameter(Table):
             ),
         }
-        super().__init__(parent, locator=locator, logger=logger, column_widgets=column_widgets)
+        self.name = list(column_widgets.keys())[0]
```
This could throw `IndexError` during instantiation if column_widgets weren't passed.

Have you checked all uses of this widget class against this requirement?

I think you'll want to make this safer by having some checking, or using try/except to set these instance attributes to `None` if column_widgets doesn't have the expected length/field format.

mshriver commented on airgun/widgets.py:
```diff
@@ -916,31 +916,44 @@ class CustomParameter(Table):
+        self.value = list(column_widgets.keys())[1]
+        self.value_key = '_'.join(self.value.lower().split(' '))
+
+        super().__init__(
```
Good work handling the kwargs above!

This call to super() isn't quite right though - no need to try and match up kwargs with the parent class, just pass the whole thing.

```
super().__init__(parent, **kwargs)
```

synkd replied:
@mshriver So I've tried changing the super() call to `super().__init__(parent, **kwargs) both in its existing location and with it moved to the first line in the this `__init__` method, but both are introducing errors in some of the other tests that use the `CustomParameter` widget. As an example, `tests/foreman/ui/test_domain.py::test_positive_set_parameter` fails with `TypeError: Table.__init__() got an unexpected keyword argument 'id'`. The same test passes with the current super() call. I'm wondering if a change to the `Table` widget would be necessary for this to work, but I'm a bit leery of making a change to such a widely-used widget. What do you think is the best approach here?

synkd replied:
So I'm still having issues with this. Making the changes that we discussed has my Ansible tests still passing, but it's failing on tests that use views that instantiate the `CustomParameter` class with an `id` argument (these are the UI test modules `test_operatingsystem.py`, `test_host.py`, and `test_domain.py`). The issue seems to be that `locator` in the table widget's `__init__` method is a positional argument with no default value. However, passing a locator argument to `super().__init__` does not resolve the issue. Instead, the Ansible test now fails, and the tests that use `CustomParameter` with `id` still fail, now because the widget seems to be evaluating to a `NoneType` object. Any ideas on how to work past this?

Here are two of the variations that I tried along with links to the tracebacks that I encountered:

```
    def __init__(self, parent, **kwargs):
        """Supports initialization via ``locator=`` or ``id=``"""
        if (
            kwargs.get('locator')
            and kwargs.get('id')
            or not kwargs.get('locator')
            and not kwargs.get('id')
        ):
            raise ValueError('Please specify either locator or id')
        locator = kwargs.get('locator') or f".//table[@id='{kwargs.pop('id')}']"
        super().__init__(parent, **kwargs)
```
results:
ansible test passes
error in host test: http://rx-paste.usersys.redhat.com/view/834c2c5f


```
    def __init__(self, parent, **kwargs):
        """Supports initialization via ``locator=`` or ``id=``"""
        if (
            kwargs.get('locator')
            and kwargs.get('id')
            or not kwargs.get('locator')
            and not kwargs.get('id')
        ):
            raise ValueError('Please specify either locator or id')
        locator = kwargs.get('locator') or f".//table[@id='{kwargs.pop('id')}']"
        super().__init__(parent, locator=locator, **kwargs)

```
results:
error in ansible test: http://rx-paste.usersys.redhat.com/view/2842dfc7
error in host test: http://rx-paste.usersys.redhat.com/view/ee7b053b

mshriver reviewed (commented):
Several non-blocking comments that you may or may not want to address in this PR.

One or two blocking comments, but nothing major.

Great work here @synkd this is a great addition to airgun!

latran reviewed (commented):
Some general comments and some clarifications.  Overall looks good.  If there's some test cases written for this, I would like to see them if possible.

synkd commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,70 @@
+    PER_PAGE_BUTTON_DROPDOWN = ".//div[button[@id='pagination-options-menu-toggle-2']]"
+
+
+class AnsibleRolesView(BaseLoggedInView, SearchableViewMixin):
```
This class represents the `Configure > Roles` page (i.e. satellite.com/ansible/ansible_roles). When no Ansible roles have been imported and this page is visited, only the import button is visible. I tried to convey this in the class's docstring. 

When writing this view, I decided that it was not necessary to have separate classes for the page representing its state before and after importing roles. The only available action on this page before importing roles is importing roles, which is available via the `import_button` widget in this class. The other widgets in this class are present after roles are imported.

synkd commented on airgun/views/ansible_role.py:
```diff
@@ -0,0 +1,70 @@
+        return self.title.is_displayed and self.import_button.is_displayed
+
+
+class AnsibleRolesImportView(BaseLoggedInView):
```
Yes, this is the `Roles > Changed Ansible Roles` page, which is loaded after the `import_button` widget in `AnsibleRolesView` is clicked.

synkd commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,75 @@
+        view.table.row()['Actions'].widget.fill('Delete')
+        view.dialog.confirm_dialog.click()
+        view.flash.assert_no_error()
+        view.flash.dismiss()
```
Ack, that is good to know. When that happens, I can work with Sam (the new Ansible component owner) to refactor this  if necessary.

synkd commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,75 @@
+
+    def delete(self, entity_name):
+        """Delete Ansible Role from Satellite"""
+        # This method currently does not work as the Name column table header
```
Good point - I will change this to reflect the reason behind doing search first, then delete. I believe the Satellite UX team is working on changing this in a future release - likely by wrapping the `▲` character in a separate `span` tag from the column header.

synkd commented on airgun/views/ansible_variable.py:
```diff
@@ -0,0 +1,116 @@
+    text input field."""
+
+    matcher_key = Select(".//select")
+    matcher_value = TextInput(locator=".//input[@class='matcher_value']")
```
Ack, in the interest of getting this PR merged, I will leave this as is for now. If a future test requirement needs to access that `delete` button, it can be added at that time.

mshriver commented on airgun/entities/ansible_role.py:
```diff
@@ -0,0 +1,79 @@
+        view = self.navigate_to(self, 'All')
+        view.search(entity_name)
+        view.table.row()['Actions'].widget.fill('Delete')
+        view.dialog.confirm_dialog.click()
```
You can now drop this line.
